### ISSUE TITLE
Cleaner handling of functors in md-flexible

### ISSUE THREAD

**Issue by SamNewcome** [AutoPas/AutoPas#796]
**Is your feature request related to a problem? Please describe.**
At the moment we have a rather spaghetti-based situation whereby the user specifies which functor they would like compiled. At the moment this includes every possible vectorisation. Then depending on whether AutoPas is in multi-site mode or not, a particular version is compiled. I see this as getting rather spaghetti-like with more and more functors - both in terms on vectorisation and the actual force model.

**Describe the solution you'd like**
A compile option `AUTOPAS_APPLICATION_LIBRARY_PREFERRED_VECTORIZATION` (or potentially a short alternative). The user sets a preferred vectorisation choice, e.g. `AUTOVEC`, `AVX512`, `SVE`. If the user's preferred choice is not available for a particular force model (LJ, Mie + Singlesite/Multisite), then a warning is thrown and a fallback option is used. For now the fallback option could be the AutoVec version, but equally, it could also be a vector wrapper version. We remove the need for vectorisation to be specified in the input.

For functors based on different force models (e.g. Lennard-Jones, Mie, (Axilrod-Teller-Muto?)), we keep the current compile time system + specification in the input. 

My suggestion is based on the assumption that users don't need to compile and use multiple different vectorisations of one force calculation to be compiled at any one time. I think that this is a fair assumption - you generally just want the best available option for your architecture. The exception to this is when you are actually comparing implementations of the same functor. But this can be easily handled with separate build folders. As this exception is pretty much just for developers of functors, I think this solution is suitable.

**Describe alternatives you've considered**
The default option is `AUTOMATIC`, which determines the best vectorisation based on architecture and what is available for a particular force model.

**Comment by FG-TUM** [comment]
We already offer the user not to compile versions of functors
https://github.com/AutoPas/AutoPas/blob/b292345510f218132b353ed775ba9c8274c1ce0c/examples/md-flexible/CMakeLists.txt#L19-22

This could (and probably should) be extended to every functor we offer in our application library.

**Comment by SamNewcome** [comment]
No, I fully understand this but my suggestion was a way of cleaning this up as new force models and vectorisation methods are developed. If this isn't clear, please let me know and I will reformulate the first comment.

The reason for the feature is that I don't see a reason why a user would want to compile multiple vectorisation types for one particular force model. In general, people just want the "best". 

My suggestion to reduce the "vectorisation" compile time options to a single option would reduce some of the options that need to be set at compile time, as well as remove the need to specify which vectorisation of, e.g. the LJ Functor, is needed at run time. Finally, having some automatic option which determines the "best" (which wouldn't be too hard to implement) would remove the decision entirely from a user or developer who just wants a good functor that runs without any modifications.

Despite this issue, I certainly don't think that the current situation is bad (at least for now) and I see the improvements from this issue as minor at best. And so I am happy to close this issue if you have a good reason for not doing this.

**Comment by FG-TUM** [comment]
Ah ok so your primary goal is to turn the manual switches into one automatic switch? I still don't fully understand whats the problem you want to solve. Is it having too many cmake options or the way we have to handle the options in cmake or cpp? 

The reasoning for the current system was that sometimes, as you stated initially, you want to compare functors for speed or qualitative reasons. This is when you want to be able to switch them on or off individually. This is btw also the way kokkos does it with its backends. 

When you just want "the best", this is where the defaults come in. If we'd also had a AVX512 Functor, we probably should implement some logic that determines if that is available which then sets the default values of the options. 

For choices like these you have to think about the general design direction you want to take. Should this stuff be more heard towards users, which means clearer, less fine grained controls, or towards developers. So far, we chose the latter. Btw. this also includes choices when to throw a warning or outright abort compilation / execution. Users are most often fine with getting the simulation result. Developers want to control what exactly happens, e. g. for a benchmark, using many well structured options (which imho isn't spaghetti). So compiling a different functor than you chose and then missing the warning, and only noticing when you look at the data or output a day later can be very frustrating. :D

I'm not against a change here, but i think it has to be done very cleanly and carefully to be a benefit.

**Comment by SamNewcome** [comment]
I think let's park the idea for now - it makes more sense to rethink this if and when this becomes an actual irritation/problem.

**Comment by SamNewcome** [comment]
Closed, as Google Highway offers a clean solution for this - for forces other than Lennard-Jones, we only plan a Google Highway explicit vectorisation, so this doesn't gain anything.
